andela/ah-robotics
# #162949086 Pagination support for Articles

### What does this PR do?
Adds pagination support for articles list view.

### Description of Task to be completed?
- [x] Add pagination tests
- [x] Add configuration for rest framework

### How should this be manually tested?
1. Open your terminal and clone this repository on your local machine using 

     `git clone https://github.com/andela/ah-robotics.git`
2. Change your working directory to the project's directory using 
     
     `cd ah-robotics`
3. Change to the specific branch `ft-article-pagination-162949086` using 

     `git checkout ft-article-pagination-162949086`
4. Create a virtual environment using the following command: 
     
     `python3 -m virtualenv env`
5. Activate the virtual environment using 
     
     `source env/bin/activate`
6. Install all the requirements using 
     
     `pip install -r requirements.txt`

    #### Database setup

    Ensure [PostgreSQL](https://www.postgresql.org/) is installed and is running, follow these steps to create a new user and a database which will be used in the application.

    1. Start the postgres database server using the command:
      
        `pg_ctl -D /usr/local/var/postgres start`
    
    2. Run psql interactive terminal
    
        `psql postgres`
    
    3. Create a new user:
    
        `CREATE USER sample_username WITH PASSWORD 'sample_password';`
    
    4. Grant privileges to the user:
    
        `ALTER USER sample_username CREATEDB;`
    
    5. Create a database:
    
        `CREATE DATABASE sample_database_name WITH OWNER sample_username;`
    
    6. Exit psql interactive with the following command: `\q`
    
7. Create a **.env** file and add the following configuration variables

    ```source-json
          source env/bin/activate
          export DB_NAME="database_name"
          export DB_USER="sample_username"
          export DB_PASS="sample_password"
          export DB_HOST="host_name_eg_localhost"
          export DB_PORT="port_number_eg_5432"
          export TIME_DELTA="30"
          export EMAIL_HOST="smtp.gmail.com"
          export EMAIL_PORT=587
          export EMAIL_HOST_USER="user@example.com"
          export EMAIL_HOST_PASSWORD="email_password"
          export PASSWORD_RESET_URL="api/v1/reset_password"
          export CLIENT_DOMAIN="127.0.0.1:8000"
          export CLOUDINARY_NAME="your_cloudinary_name"
          export CLOUDINARY_KEY="your_cloudinary_key"
          export CLOUDINARY_SECRET="your_secret_key"
    ```
    
    > **Note**: The values of some of these variables are private namely the EMAIL_HOST_USER, EMAIL_HOST_PASSWORD and the CLOUDINARY VALUES.  Use your GMAIL email address and password for the  EMAIL_HOST_USER and EMAIL_HOST_PASSWORD respectively. To get cloudinary values, create an account with cloudinary and obtain the values.

8. Source .env to set the configuration variables in the terminal:

      `source .env`
9. Run migrations through the terminal to create the tables in the database:
      
      (i) Make the migrations using: 
      
      `python manage.py makemigrations`
      
      (ii) Apply the migrations using: 
      
      `python manage.py migrate`

10. Run the following command to collect static files.

    `python manage.py collectstatic`
      
#### Testing the features

1. Run the server from the terminal using: 
    
    `python manage.py runserver`
2. Ensure [Postman](https://www.getpostman.com/) is installed, open it on your local machine and Paste the URL `http://localhost:8000/api/v1/users/` to first sign up a user.
3. On the **Headers** section add **Content-type** as the KEY and **_application/json_** as the VALUE.
4. Head over to the **Body** section and select the **raw** option and add the following payload to the body:

    ```
        {
            "user": {
                "username": "sample_username",
                "email": "sample_email",
                "password": "sample_password"
                }
        }
    ```
5. Once all the fields are inserted, ensure that it is a **POST** request and click the **Send** button.
6. Log in the user using the following endpoint `http://localhost:8000/api/v1/users/login/` and insert the following payload to the body section.
    
    ```
        {
            "user": {
                "email": "sample_email",
                "password": "sample_password"
                }
        }
    ```
    Upon successful login, a token is returned to the user, this token will be used to access the protected endpoints
7. Create several articles, insert the following URL: `http://localhost:8000/api/v1/articles/`. In the body section, ensure the type is _raw_ and the format is _JSON(application/json)_. Finally, post an article by using the following payload:
    
    ```
    {
        "article": {
            "title": "Sample Article",
            "description": "Sample Article Description",
            "body": "The sample article body",
            "tagList": ["samples", "articles"]
        }
    }
    ```
    Once all the fields are inserted, change request type to a **POST** request and repeatedly click the **Send** button as many as 20 to 30 times.

9. To see the pagination effect in action, head over to `http://localhost:8000/api/v1/articles/` and perform a **GET** request. In the response body, you'll be able to see the  *count* value which represents the total number of articles, a **next** value which represents the link to the next page and *previous* which is a link to the previous page. 
    > The **next** and **previous** values are null by default if the page limit is not exceeded.

10. To load the second page of articles, click on the value of  **"next"** key in the response body,  the value in question should be a URL that will load the second page of paginated data, once clicked, the URL will be automatically added to the postman URL section, change the request type to **Get** and then click **Send** 
    > The URL should be similar to `http://localhost:8000/api/v1/articles/?page=2`

#### Running tests
1. Head over to the terminal and ensure that the server is not running and you are in the project root folder.
2. Run the tests by typing in the following command: 
    
    `python manage.py test`

### Any background context you want to provide?
None
### What are the relevant pivotal tracker stories?
[#162949086](https://www.pivotaltracker.com/n/projects/2235171/stories/162949086)

### Screenshots
1. load articles list which has not exceeded page limit number
    <img width="1440" alt="screen shot 2019-02-05 at 8 07 47 pm" src="https://user-images.githubusercontent.com/10683903/52292138-19197a00-2985-11e9-93b8-309245f49281.png">


2. after adding more articles in order to exceed page limit
    <img width="1440" alt="screen shot 2019-02-05 at 8 14 10 pm" src="https://user-images.githubusercontent.com/10683903/52292223-4d8d3600-2985-11e9-8941-2f394b26066b.png">

2. loading the second page
    <img width="1440" alt="screen shot 2019-02-05 at 8 14 42 pm" src="https://user-images.githubusercontent.com/10683903/52292402-b1affa00-2985-11e9-915b-6d5c04584afb.png">

3. loading the third page
    <img width="1440" alt="screen shot 2019-02-05 at 8 14 56 pm" src="https://user-images.githubusercontent.com/10683903/52292428-be345280-2985-11e9-8082-6fe7781902ab.png">

## Questions:
Questions and comments are welcome.
